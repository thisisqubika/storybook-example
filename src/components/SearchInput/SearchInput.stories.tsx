import type { Meta, StoryObj } from '@storybook/react';
import { expect, fn, userEvent, within } from '@storybook/test';
import { ThemeWrapper } from '../../common/themes';
import { SearchInput, type SearchInputProps } from '.';

export default {
  component: SearchInput,
} satisfies Meta<SearchInputProps>; // meta

export const Base: StoryObj<SearchInputProps> = {
  args: {
    color: 'black',
    disabled: false,
    initialTerm: '',
    onSearch: fn(),
  } satisfies SearchInputProps,

  render(props: SearchInputProps) {
    return (
      <ThemeWrapper>
        <SearchInput {...props} />
      </ThemeWrapper>
    );
  }, // render

  async play({ args, canvasElement }) {
    const canvas = within(canvasElement);
    const searchInput = canvas.getByTestId('SearchInput.input');
    const testInputText = 'Book search';
    await userEvent.type(searchInput, testInputText, { delay: 100 });
    const iconButton = canvas.getByRole('button');
    await userEvent.click(iconButton);
    await expect(args.onSearch).toHaveBeenCalledWith(testInputText);
  }, // play

}; // const Base
