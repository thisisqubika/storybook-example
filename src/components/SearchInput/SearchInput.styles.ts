import { SxProps, Theme, alpha, styled } from "@mui/material";

export const Search = styled('div')(({ theme }) => ({
  alignItems: 'center',
  borderRadius: theme.shape.borderRadius,
  backgroundColor: alpha(theme.palette.common.white, 0.15),
  '&:hover': {
    backgroundColor: alpha(theme.palette.common.white, 0.25),
  },
  width: '100%',
  [theme.breakpoints.up('sm')]: {
    marginLeft: theme.spacing(1),
    width: 'auto',
  },
})); // const Search

export function sxStyles(): Record<string, SxProps<Theme>> {
  return {
    textField: {
      // color: 'white',
      width: '100%',
    }, // textField
  };
}; // sxStyles
