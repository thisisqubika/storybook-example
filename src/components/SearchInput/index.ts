export {
  default as SearchInput,
  type SearchInputProps,
} from './SearchInput';
