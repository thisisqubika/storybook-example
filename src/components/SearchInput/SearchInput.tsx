import { CSSProperties, useState } from "react";
import SearchIcon from '@mui/icons-material/Search';
import { Search, sxStyles } from "./SearchInput.styles";
import { IconButton, InputAdornment, TextField } from "@mui/material";

export interface SearchInputProps {
  color?: CSSProperties['color'];
  disabled?: boolean;
  initialTerm?: string;
  onSearch: (term: string) => void;
} // interface SearchInputProps

export default function SearchInput({
  color,
  disabled = false,
  initialTerm = '',
  onSearch,
}: SearchInputProps) {
  const styles = sxStyles();
  const [term, setTerm] = useState(initialTerm);

  const endAdornment = (
    <InputAdornment position='end'>
      <IconButton
        data-testid='SearchInput.IconButton'
        onClick={() => onSearch(term)}
      >
        <SearchIcon sx={{ color }} />
      </IconButton>
    </InputAdornment>
  ); // const endAdornment

  return (
    <Search>
      <TextField
        data-testid='SearchInput'
        disabled={disabled}
        inputProps={{ 'data-testid': 'SearchInput.input' }}
        InputProps={{
          endAdornment,
          sx: { color },
        }}
        onChange={(e) => setTerm(e.target.value)}
        onKeyUp={(e) => {
          if (e.key === 'Enter') onSearch(term);
        }}
        placeholder='Search...'
        sx={styles.textField}
        value={term}
        variant="standard"
      />
    </Search>
  );
} // function SearchInput
