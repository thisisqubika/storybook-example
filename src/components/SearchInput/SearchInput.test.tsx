import { describe, expect, test } from 'vitest';
import { composeStories } from "@storybook/react";
import { render } from '@testing-library/react';
import * as stories from './SearchInput.stories';
import { type SearchInputProps } from './SearchInput';

const {
  Base: SearchInputBaseStory,
} = composeStories(stories);

describe('SearchInput', () => {
  test('stories match snapshot', () => {
    for (const Story of [
      SearchInputBaseStory,
    ]) {
      const props = Story.args as SearchInputProps;
      const { container } = render(
        <Story {...props} />
      );
      expect(container).toMatchSnapshot();
    }
  }); // test

}); // describe 'SearchInput'
