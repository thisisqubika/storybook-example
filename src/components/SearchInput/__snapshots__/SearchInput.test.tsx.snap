// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchInput > stories match snapshot 1`] = `
<div>
  <div
    class="css-r7zuen"
  >
    <div
      class="MuiFormControl-root MuiTextField-root css-136h9lv-MuiFormControl-root-MuiTextField-root"
      data-testid="SearchInput"
    >
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-1lohisl-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd css-1x51dt5-MuiInputBase-input-MuiInput-input"
          data-testid="SearchInput.input"
          id=":r0:"
          placeholder="Search..."
          type="text"
          value=""
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1m1cnyu-MuiButtonBase-root-MuiIconButton-root"
            data-testid="SearchInput.IconButton"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-xhi8ul-MuiSvgIcon-root"
              data-testid="SearchIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
